clusterloader2/pkg/measurement/measurement_executor.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package measurement
+
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/perf-tests/clusterloader2/api"
+	"k8s.io/perf-tests/clusterloader2/pkg/errors"
+)
+
+// Execute executes a measurement, which can be a single measurement or a wrapper for multiple measurements.
+// It throws an error if both identifier and Instances has been supplied in the config.
+func Execute(mm MeasurementManager, m *api.Measurement) error {
+	if len(m.Instances) != 0 && m.Identifier != "" {
+		return fmt.Errorf("only one of instances or identifier must be supplied. Measurement method - %s, identifier - %s, instances - %v", m.Identifier, m.Method, m.Instances)
+	}
+	if m.Identifier != "" {
+		return mm.Execute(m.Method, m.Identifier, m.Params)
+	}
+	return executeWrapperMeasurement(mm, m)
+}
+
+func executeWrapperMeasurement(mm MeasurementManager, m *api.Measurement) error {
+	var wg wait.Group
+	errList := errors.NewErrorList()
+	for i := range m.Instances {
+		identifier := m.Instances[i].Identifier
+		measurementInstanceParams := m.Instances[i].Params
+		measurementParams := m.Params
```

Comment:
I think we need to create a copy of m.Params here in each iteration. Otherwise it won't work correctly in the following scenario:
```
Measurement:
   Params:
      C: true
   Instances:
      -Identifier: A
       Params:
           A: true
      -Indentifier: B
       Params:
           B: true
```

In such case I'd expect measurement A to be executed with params `A` and `C` set and B to be executed with params `B` and `C`. But, your code will execute B with `A`, `B` and `C` set

```

Reply:
good catch

@vamossagar12 - please chang

Reply:
hey @mm4tt slight confusion here. `measurementParams := m.Params` isn't it already creating a local copy of m.params? am i missing something here?

And on a side note, is there a way to write some basic unit tests in perf-tests? I see there are some tests but can those be used? I say because this particular method has had some back and forth(could also be because of my limited understanding of golang since i haven't worked much with it). But these cases sometimes can be caught by unit-tests

Reply:
Yeah, if you have a capacity to add a unit test it would be great.
To create a unit test, you just need to add a file ending with _test.go, here is an example - https://github.com/kubernetes/perf-tests/blob/996cd6d852ce8c8188bdff9f6abb58609763e2bf/clusterloader2/pkg/prometheus/util_test.go

And to answer the original question, the map type in golang is always a reference type. So a line like 
` measurementParams := m.Params`

doesn't copy the map, just creates another reference pointing to the same object.
If you modify `measurementParams` you'll also modify the `m.Params`.

You should do something like:
```
measurementParams = make(map[string]interface{})
for k,v := range m.Params {
  measurementParams[k] = v
}
```

Reply:
thanks @mm4tt henceforth would write unit tests wherever there's a doubt. i made the change you described.